components/layout_2020/display_list/stacking_context.rs

```diff
@@ -500,57 +549,90 @@ impl StackingContext {
     pub(crate) fn build_display_list(&self, builder: &mut DisplayListBuilder) {
         let pushed_context = self.push_webrender_stacking_context_if_necessary(builder);
 
-        // Properly order display items that make up a stacking context. "Steps" here
-        // refer to the steps in CSS 2.1 Appendix E.
+        #[cfg(debug_assertions)]
+        {
+            let indent = "    ".repeat(builder.debug_print_indent);
+            match self.context_type {
+                StackingContextType::Real => {
+                    debug!("{}{:?} z={}", indent, self.context_type, self.z_index())
+                },
+                _ => debug!("{}{:?}", indent, self.context_type),
+            }
+            builder.debug_print_indent += 1;
+        }
+
+        // Properly order display items that make up a stacking context.
+        // “Steps” here refer to the steps in CSS 2.1 Appendix E.
+        // Note that “positioned descendants” is generalised to include all descendants that
+        // generate stacking contexts (csswg-drafts#2717), except in the phrase “any positioned
+        // descendants or descendants that actually create a stacking context”, where the term
+        // means positioned descendants that do not generate stacking contexts.
 
         // Steps 1 and 2: Borders and background for the root
         let mut child_fragments = self.fragments.iter().peekable();
         while child_fragments.peek().map_or(false, |child| {
-            child.section == StackingContextSection::BackgroundsAndBorders
+            child.section() == StackingContextSection::BackgroundsAndBorders
         }) {
-            child_fragments.next().unwrap().build_display_list(builder);
+            child_fragments
+                .next()
+                .unwrap()
+                .build_display_list(builder, &self.inline_stacking_containers);
         }
 
-        // Step 3: Positioned descendants with negative z-indices
-        let mut child_stacking_contexts = self.stacking_contexts.iter().peekable();
-        while child_stacking_contexts
+        // Step 3: Stacking contexts with negative ‘z-index’
```

Comment:
I may be wrong, but I think the only stacking contexts with negative effective z-indices are those that are positioned. Not sure if this deserves more clarification.

Reply:
i think you’re right for now, since flex and grid aren’t yet implemented, but ‘z-index’ applies to [flex items](https://drafts.csswg.org/css-flexbox/#painting) and [grid items](https://drafts.csswg.org/css-grid/#z-order) even when ‘position’ is ‘static’. (that said, it’s not clear to me whether “behaves exactly as if ‘position’ were ‘relative’” is limited to ‘z-index’, or if the used value becomes ‘relative’ in every sense…)